crates/sui-core/src/execution_engine.rs

```diff
@@ -490,14 +490,14 @@ fn pay<S>(
                 DeleteKind::Normal,
             );
         } else {
-            // unwrapped unsafe because we checked that it was a coin object above
-            coin_object
-                .data
-                .try_as_move_mut()
-                .unwrap()
-                .update_contents_and_increment_version(
-                    bcs::to_bytes(&coin).expect("Coin serialization should not fail"),
-                );
+            let new_contents = bcs::to_bytes(&coin).expect("Coin serialization should not fail");
+            // unwrap safe because we checked that it was a coin object above
+            let move_obj = coin_object.data.try_as_move_mut().unwrap();
+            debug_assert_eq!(move_obj.type_specific_contents().len(), new_contents.len());
```

Comment:
we should either use assert! or move this check.

I'm curious if `1u16` and `65535u16` takes the same amount of space in its representation?

Reply:
Yep, every `u16` will take 16 bytes no matter what is value is. BCS only compresses integers that represent vector lengths.

Reply:
I decided to get rid of these asserts--they are redundant with what `update_contents_and_increment_version` is checking.